### ISSUE TITLE
Cursor not found when poolSize > 1

### ISSUE THREAD

**Issue by iamdoron** [christkv/mongodb-core#89]
I'm not sure if it's a driver issue. I'm connecting to mongos (2 servers) through ssl with the following configuration:

```
const mongoOptions = {
 mongos: {
    ssl: true,
    sslValidate: true,
    sslCA: ca,
    poolSize: <poolSize>
  },
  acceptableLatencyMS: 100,
  socketOptions: {
    keepAlive: 120
  }
}
```

mongodb-driver version 2.1.12, mongodb-core 1.3.12

I'm running multiple queries (using the driver find().toArray()); 4 return empty results, 2 return ~1k results out of ~30k in the whole collection, the ~1k query takes  ~400ms to complete.

sometimes I get the following errors when poolSize > 1
```
{ [MongoError: Cursor not found (namespace: 'db.collection', id: 9047773609076924875).]
  name: 'MongoError',
  message: 'Cursor not found (namespace: \'db.collection\', id: 9047773609076924875).',
  ok: 0,
  errmsg: 'Cursor not found (namespace: \'db.collection\', id: 9047773609076924875).',
  code: 43 }
``` 

The more you increase the poolSize, the higher the chance of it happening. And, of course, increasing the number of queries would also increase the error rate.
with poolSize=1 I wasn't able to reproduce this issue

**Comment by christkv** [comment]
Hmm can you create a repo script for this. Cursors are bound to a specific proxy for the duration of the life cycle. This error means whatever mongod the mongos is forwarding to is not finding the cursor with that id which is weird. Are the collections you are querying sharded ?

What version of mongod and mongos are you running ?

**Comment by iamdoron** [comment]
3.2.1, the collection is not sharded

I'll create a script that reproduces it :+1:

**Comment by iamdoron** [comment]
using poolSize = 10

```javascript
const collection = db.collection("tempCollection");
  collection.remove({}, (err) => {
    if (err) return console.trace(err);
    collection.createIndexes([{key: {number: 1}}], (err) => {
      const documents = [];
      for (var i = 0; i < 30000; i++) {
        documents.push({number: i % 15, prop: '+ftnmrzxEcb+0wCcZjC/YWKP5K0M33FC+FyTaPKxCbyEkFWmXgdU+QW7I6HxDmUPjN8='})
      }
      collection.insertMany(documents, (err) => {
        if (err) return console.trace(err);
        console.log("finish inserting documents...");
        collection.find({number: 2}).toArray((err, results) => {
          if (err) return console.trace(err);
          console.log("got", results.length, "results");
        })
        collection.find({number: 2}).toArray((err, results) => {
          if (err) return console.trace(err);
          console.log("got", results.length, "results");
        })
        collection.find({number: 2}).toArray((err, results) => {
          if (err) return console.trace(err);
          console.log("got", results.length, "results");
        })
        collection.find({number: 2}).toArray((err, results) => {
          if (err) return console.trace(err);
          console.log("got", results.length, "results");
        })
        collection.find({number: 2}).toArray((err, results) => {
          if (err) return console.trace(err);
          console.log("got", results.length, "results");
        })
        collection.find({number: 2}).toArray((err, results) => {
          if (err) return console.trace(err);
          console.log("got", results.length, "results");
        })
        collection.find({number: 2}).toArray((err, results) => {
          if (err) return console.trace(err);
          console.log("got", results.length, "results");
        })
        collection.find({number: 2}).toArray((err, results) => {
          if (err) return console.trace(err);
          console.log("got", results.length, "results");
        })
        collection.find({number: 3000}).toArray((err, results) => {
          if (err) return console.trace(err);
          console.log("got", results.length, "results");
        })
        collection.find({number: 3000}).toArray((err, results) => {
          if (err) return console.trace(err);
          console.log("got", results.length, "results");
        })
        collection.find({number: 3000}).toArray((err, results) => {
          if (err) return console.trace(err);
          console.log("got", results.length, "results");
        })
      })
    })
```

**Comment by christkv** [comment]
can't get it to happen, is it possible for you to provide a trace of where the error is originating ?

**Comment by iamdoron** [comment]
```
Trace: { [MongoError: Cursor not found (namespace: 'db.collection', id: 2097333442636404359).]
  name: 'MongoError',
  message: 'Cursor not found (namespace: \'db.collection\', id: 2097333442636404359).',
  ok: 0,
  errmsg: 'Cursor not found (namespace: \'db.collection\', id: 2097333442636404359).',
  code: 43 }
    at repl:31:25
    at handleCallback (server/node_modules/mongodb/lib/utils.js:96:12)
    at server/node_modules/mongodb/lib/cursor.js:832:22
    at handleCallback (server/node_modules/mongodb-core/lib/cursor.js:159:5)
    at server/node_modules/mongodb-core/lib/cursor.js:580:24
    at queryCallback (server/node_modules/mongodb-core/lib/wireprotocol/3_2_support.js:183:14)
    at bound (domain.js:280:14)
    at runBound (domain.js:293:12)
    at Callbacks.emit (server/node_modules/mongodb-core/lib/topologies/server.js:116:3)
    at null.messageHandler (server/node_modules/mongodb-core/lib/topologies/server.js:291:23)
    at TLSSocket.<anonymous> (server/node_modules/mongodb-core/lib/connection/connection.js:285:22)
    at emitOne (events.js:77:13)
    at TLSSocket.emit (events.js:169:7)
    at readableAddChunk (_stream_readable.js:146:16)
    at TLSSocket.Readable.push (_stream_readable.js:110:10)
    at TLSWrap.onread (net.js:523:20)
```

**Comment by christkv** [comment]
Hmm weird cursor is gone which is unexpected when a getMore returns a cursorId. I think I have a fix and will publish a new mongodb-core if you can test for me.

**Comment by iamdoron** [comment]
of course, I will test it

**Comment by christkv** [comment]
The refactoring of the pool into a worker queue detached the getMore calls from a specific connection which seems to cause some issues with mongos since connections are pinned all the way through. However it's weird to see this in a situation where the collection is not sharded.

**Comment by iamdoron** [comment]
I tried to look back at the logs, and I found similar issue from mongodb-driver v2.1.7 (before I upgraded). 

```
at Function.MongoError.create (server/node_modules/mongodb-core/lib/error.js:31:11)
    at queryCallback (server/node_modules/mongodb-core/lib/wireprotocol/3_2_support.js:183:34)
    at Callbacks.emit (server/node_modules/mongodb-core/lib/topologies/server.js:116:3)
    at null.messageHandler (server/node_modules/mongodb-core/lib/topologies/server.js:282:23)
    at TLSSocket.<anonymous> (server/node_modules/mongodb-core/lib/connection/connection.js:273:22)
    at emitOne (events.js:77:13)
    at TLSSocket.emit (events.js:169:7)
    at readableAddChunk (_stream_readable.js:146:16)
    at TLSSocket.Readable.push (_stream_readable.js:110:10)
    at TLSWrap.onread (net.js:523:20)
```

**Comment by christkv** [comment]
alright, just manually update the mongodb driver package to use mongodb-core 1.3.13 and let me know if the issue is resolved. if it works I'll release a mongodb version update. Thanks for helping chase down these issues caused by the pool refactoring.

**Comment by iamdoron** [comment]
Thanks! I ran it more than 10 times with no issues. I tried heavier loads and different poolSize settings

**Comment by christkv** [comment]
2.1.14 published to NPM

**Comment by christkv** [comment]
Hi I'm making a small alteration to the cursor code for the next release and might ask you to just confirm it does not break things if you are willing ?

**Comment by iamdoron** [comment]
yes 👍

**Comment by christkv** [comment]
Ok so master still working fine for you ?

**Comment by iamdoron** [comment]
just to make sure I installed 
```
npm i mongodb-core@1.3.12
```
saw the error again when running the script above

and then I installed
```
npm i git+ssh@github.com:christkv/mongodb-core.git
```

saw no errors  (tried 5 times)

**Comment by YouriT** [comment]
@christkv could you update `mongodb-native` `2.2` branch to reflect this change?
I see that `mongodb-native: 2.2.x` depends on `mongodb-core: 2.1.13`

**Comment by YouriT** [comment]
Is it possible that this exact same issue is happening using replicaset and not a mongos proxy? Because I just noticed this exact same error:
```
MongoError: Cursor not found, cursor id: 41923610084
    at Function.MongoError.create (/app/node_modules/mongodb-core/lib/error.js:31:11)
    at /app/node_modules/mongodb-core/lib/connection/pool.js:497:72
    at authenticateStragglers (/app/node_modules/mongodb-core/lib/connection/pool.js:443:16)
    at Connection.messageHandler (/app/node_modules/mongodb-core/lib/connection/pool.js:477:5)
    at Socket.<anonymous> (/app/node_modules/mongodb-core/lib/connection/connection.js:321:22)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at TCP.onread (net.js:547:20)
```
But we are using a replicaset also with a poolSize of 10.

**Comment by iamdoron** [comment]
I know it doesn't help for your issue, but what he meant was that  https://github.com/mongodb/node-mongodb-native was published to 2.1.14 while mongodb-core was published as 1.3.13
https://github.com/mongodb/node-mongodb-native/commit/f8dbce4be484aaa222464d352dab1a5c157a7098
it was long time ago :)

**Comment by YouriT** [comment]
just notice indeed I mixed up the dates my bad
